staging/src/k8s.io/apiserver/pkg/registry/rest/delete.go

```diff
@@ -200,3 +200,23 @@ func AdmissionToValidateObjectDeleteFunc(admit admission.Interface, staticAttrib
 		return nil
 	}
 }
+
+// WithCorruptObjDeleter wraps the given rest.GracefulDeleter that is used for normal
+// deletion flow with an unsafe deleter to facilitate deletion of corrupt object
+func WithCorruptObjDeleter(deleter GracefulDeleter, opts *metav1.DeleteOptions) GracefulDeleter {
+	if opts == nil {
+		return deleter
+	}
+	if ignore := opts.IgnoreStoreReadErrorWithClusterBreakingPotential; ignore == nil || !*ignore {
+		return deleter
+	}
```

Comment:
all this is not needed if you inline the code below directly into `DeleteResource`:
```golang
		if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
			if ptr.Deref(options.IgnoreStoreReadErrorWithClusterBreakingPotential, false) {
				if p, ok := r.(rest.CorruptObjectDeleterProvider); !ok {
					scope.err(fmt.Errorf("ignoreStoreReadErrorWithClusterBreakingPotential is not supported"), w, req)
					return
				} else {
					r = p.GetCorruptObjDeleter()
				}
			}
		}
```
Or in other words: pulling logic apart and making it apparently reusable makes it much easier to reason about. This function here is a noop in certain case, but that's totally invisible from the function name and signature.

Has it been called out (maybe in the KEP) that we ignore the options flags for storages that don't support it?

Reply:
the KEP does not mention this, it's a developer error if the feature is enabled, and `r.(rest.CorruptObjectDeleterProvider)` fails.